cmd/frontendapi/apisrv/apisrv.go

```diff
@@ -152,149 +130,107 @@ func (s *frontendAPI) DeleteRequest(c context.Context, g *frontend.Group) (*fron
-// the 'connstring' field of that key once it exists on the channel.
-//
-// The pattern for this function is from 'Go Concurrency Patterns', it is a function
-// that wraps a closure goroutine, and returns a channel.
-// reference: https://talks.golang.org/2012/concurrency.slide#25
-func (s *frontendAPI) watcher(ctx context.Context, pool *redis.Pool, key string) <-chan string {
-	// Add the key as a field to all logs for the execution of this function.
-	feLog = feLog.WithFields(log.Fields{"key": key})
-	feLog.Debug("Watching key in statestorage for changes")
+// GetUpdates is this service's implementation of the GetUpdates gRPC method defined in frontend.proto
+func (s *frontendAPI) GetUpdates(p *frontend.Player, assignmentStream frontend.Frontend_GetUpdatesServer) error {
+	// Get cancellable context
+	ctx, cancel := context.WithCancel(assignmentStream.Context())
+	defer cancel()
 
-	watchChan := make(chan string)
+	// Create context for tagging OpenCensus metrics.
+	funcName := "GetAssignment"
+	fnCtx, _ := tag.New(ctx, tag.Insert(KeyMethod, funcName))
 
-	go func() {
-		// var declaration
-		var results string
-		var err = errors.New("haven't queried Redis yet")
-
-		// Loop, querying redis until this key has a value
-		for err != nil {
-			select {
-			case <-ctx.Done():
-				// Cleanup
-				close(watchChan)
-				return
-			default:
-				results, err = s.retrieveConnstring(ctx, pool, key, s.cfg.GetString("jsonkeys.connstring"))
-				if err != nil {
-					time.Sleep(5 * time.Second) // TODO: exp bo + jitter
-				}
-			}
+	// get and return connection string
+	watchChan := redispb.PlayerWatcher(ctx, s.pool, *p) // watcher() runs the appropriate Redis commands.
+	timeoutChan := time.After(time.Duration(s.cfg.GetInt("api.frontend.timeout")) * time.Second)
+
+	for {
+
+		select {
+		case <-ctx.Done():
+			// Context cancelled
+			feLog.WithFields(log.Fields{
+				"playerid": p.Id,
+			}).Info("client closed connection successfully")
+			stats.Record(fnCtx, FeGrpcRequests.M(1))
+			return nil
+		case <-timeoutChan: // Timeout reached without client closing connection
+			// TODO:deal with the fallout
+			err := errors.New("server timeout reached without client closing connection")
+			feLog.WithFields(log.Fields{
+				"error":     err.Error(),
+				"component": "statestorage",
+				"playerid":  p.Id,
+			}).Error("State storage error")
+
+			// Count errors for metrics
+			errTag, _ := tag.NewKey("errtype")
+			fnCtx, _ := tag.New(ctx, tag.Insert(errTag, "watch_timeout"))
+			stats.Record(fnCtx, FeGrpcErrors.M(1))
+			return err
+
+		case a := <-watchChan:
+			feLog.WithFields(log.Fields{
+				"assignment": a.Assignment,
+				"playerid":   a.Id,
+				"status":     a.Status,
+				"error":      a.Error,
+			}).Info("updating client")
+			assignmentStream.Send(&a)
+			stats.Record(fnCtx, FeGrpcStreamedResponses.M(1))
+			// Reset timeout.
+			timeoutChan = time.After(time.Duration(s.cfg.GetInt("api.frontend.timeout")) * time.Second)
```

Comment:
Behavior callout. Overuse of the status field could mean infinitely connected clients. Do we want an actual _disconnect players after x minutes no matter what_ timeout in addition to the _touched_ timeout?

Reply:
It's kind of a protect you from yourself sort of thing, so maybe not.